MIP/mip-61/README.md

```diff
@@ -0,0 +1,194 @@
+
+- `Initiate_transfer` on the source chain (called by user)
+- `Complete_transfer` on the target chain (called by relayer)
+
+This algo is compatible with using finalized blocks (see the remark on L2 complete verification). By pulling only finalized source blocks, the  nonce are always updated on finalization.
+
+In this MIP we consider the lock/mint Native Bridge, see [MIP-60](https://github.com/movementlabsxyz/MIP/pull/60). We will discuss a single direction of transfer - from **source chain** to **target chain**.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+!!! . The algorithm is the same in both directions. Hence each algorithm has to be implemented twice. Once for the transfer direction L1 --> L2, and once for the L2 --> L1 direction.
+
+The relayer can get its state from the source chain and target chain. The algo has three parts:
+
+**1. Continuous-Processing**: The relayer pulls transfer events from the source chain and processes them. For each transfer the relayer has to check if the transfer is completed on the target chain, and if not it has to complete the transfer.
+
+**2. Bootstrap:** The relayer initiate the event polling at some point in the past. This point is either the genesis, some configurable source block height, or some checkpoint (possibly on-chain)
+
+#### Event-Processing
+
+The relayer manages only one state of transfer: pending, which exists during the time between the reception of the `initiate_transfer` event from the source chain and the success of the `complete_transfer` tx on the target chain.
+
+The Native Bridge protocol MUST implement the assignment on the source chain of an incrementing `nonce`. Since there are two directions there MUST be two counters - one for each direction. The `nonce` is used to order the transfers.
+
+For each `initiate_transfer` event do the following:
+
+![alt text](event_processing.png)
+
+1. Get the `nonce`.
+1. If a nonce has been skipped this should not happen and we should report / throw an error.
+1. If nonce already is recorded check the `status_transfer`.
+    1. If completed set status transfer as `transfer_completed` and return.
+    1. Else create `nonce` entry locally and query target chain if `complete_transfer` transaction has been submitted.
+        1. If `complete_transfer` transaction is on target chain
+            1. If the transaction is final, set `transfer_complete` for the status of the `nonce`
+            1. Else set status us `transfer_pending` and set a timeout, for the relayer to check later again.
+        1. Else 
+            1. send `complete_transfer` transaction to target chain
+            1. set the status as `transfer_pending`.
+            1. set a timeout after which the timeout-process should trigger for the transfer.
```

Comment:
>  I think there's no easy way to know if a Tx has been sent for the complete call and not executed. 

just to not misunderstand. There is a difference between a transaction been executed, and a transaction beeing finalized completed. We are not interested in in-flight transactions. But we can differentiate between transactions that are succesfully executed on the chain (e.g. L2) and those that have also passed the finality criteria. Now the finality critera on L2 could be either single blocks already or with different consensus multiple blocks.. it also could be after MCR (i.e. Postconfirmation on L1) have protected the L2 state which would be 30minutes for L2.

Reply:
> Query L2 processed nonce. If present ...

I state which chain is querried by `target` and `source` . the algorithm is symmetrical so we need to avoid saying L1 or L2 here. But you are correct, i will double check that i use `target` and `source` everywhere.

Reply:
I say that because I understand the following statement:
> Else create nonce entry locally and query target chain if complete_transfer transaction has been submitted.

As I've to query the chain for submitted complete_transfer tx that is not really possible. Only final Tx can be used. That why I say that we should change these sentence to onchain state query and remove all transaction verification sentence.

Reply:
Yes you're right, we should use `target` and `source`.

Reply:
My remarks suppose the relayer don't know the complete_tranfer tx. I suppose an admin can execute a complete_transfer Tx. The relayer can't easily query the chain to detect this Tx before it's finalized.
So for a nonce the relayer can't be sure there's no pending Tx for it.

Reply:
can we get two states from a node? That is, the k-confirmed state and the finalized state? Worst case can we run two nodes with different configurations for this?